### PR TITLE
Dev/cya/add duedate task

### PR DESCRIPTION
**Pull request by ClayAchahui**
GitHub Issue (If applicable): #

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

We can not add due date or modify it

## What is the new behavior?

We can set due date(is still missing with "pick a date")

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/ToDo.UI/ReactiveClasses.cs [hunk 1]
```diff
@@ -17,7 +17,7 @@ public static partial class ReactiveViewModelMappings
 				{ typeof(SettingsViewModel),typeof(SettingsViewModel.BindableSettingsViewModel)},
 				{ typeof(WelcomeViewModel),typeof(WelcomeViewModel.BindableWelcomeViewModel)},
 				{ typeof(TaskListViewModel),typeof(TaskListViewModel.BindableTaskListViewModel )},
-				{ typeof(TaskViewModel),typeof(TaskViewModel.BindableTaskViewModel)}
+				{ typeof(TaskViewModel),typeof(TaskViewModel.BindableTaskViewModel)},
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/ReactiveClasses.cs:20]: This Change is not needed here I think

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml [hunk 2]
```diff
@@ -13,8 +13,77 @@
 	  Background="{ThemeResource BackgroundBrush}">
 
 	<Page.Resources>
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:15]: Did you generate the code with the proper last version of the Material M3 file ?

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml [hunk 3]
```diff
@@ -13,8 +13,77 @@
...
-		<x:String x:Key="Icon_Chevron_Right">F1 M 1.4099998474121094 0 L 0 1.4099998474121094 L 4.579999923706055 6 L 0 10.59000015258789 L 1.4099998474121094 12 L 7.409999847412109 6 L 1.4099998474121094 0 Z</x:String>
+		<ResourceDictionary>
+			<x:String x:Key="Icon_Add">F1 M 14 8 L 8 8 L 8 14 L 6 14 L 6 8 L 0 8 L 0 6 L 6 6 L 6 0 L 8 0 L 8 6 L 14 6 L 14 8 Z</x:String>
+			<x:String x:Key="Icon_Calendar_Today">F1 M 18 2 L 17 2 L 17 0 L 15 0 L 15 2 L 5 2 L 5 0 L 3 0 L 3 2 L 2 2 C 0.8999999761581421 2 0 2.899999976158142 0 4 L 0 20 C 0 21.100000023841858 0.8999999761581421 22 2 22 L 18 22 C 19.100000023841858 22 20 21.100000023841858 20 20 L 20 4 C 20 2.899999976158142 19.100000023841858 2 18 2 Z M 18 20 L 2 20 L 2 7 L 18 7 L 18 20 Z</x:String>
+			<x:String x:Key="Icon_Chevron_Right">F1 M 1.4099998474121094 0 L 0 1.4099998474121094 L 4.579999923706055 6 L 0 10.59000015258789 L 1.4099998474121094 12 L 7.409999847412109 6 L 1.4099998474121094 0 Z</x:String>
+
+			<Style x:Key="MyCustomDatePickerStyle"
```
Comment (Xiaoy312) [inline src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:21]: any particular reason why a custom style was necessary?
and if it is really needed, try to name it accordingly to its function

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml [hunk 4]
```diff
@@ -13,8 +13,77 @@
...
+										 <!--PlaceholderText--> 
+										<TextBlock x:Name="PlaceholderText"
+											   Margin="0,0,0,0"
+										   HorizontalAlignment="Left"
+										   VerticalAlignment="Center"
+										   Grid.Column="1"
+											FontWeight="Bold"
+											   Text="{TemplateBinding Header}"/>
+
+										<!--<Path Data="M9.99984 6L8.58984 7.41L13.1698 12L8.58984 16.59L9.99984 18L15.9998 12L9.99984 6Z"
```
Comment (Xiaoy312) [inline src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:56]: remove?

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml [hunk 5]
```diff
@@ -13,8 +13,77 @@
...
+					TargetType="DatePicker">
+				<Setter Property="Template">
+					<Setter.Value>
+						<ControlTemplate TargetType="DatePicker">
+							<Grid x:Name="LayoutRoot">
+								<!-- Flyout Button -->
+								<Button x:Name="FlyoutButton"
+								HorizontalAlignment="Stretch"
+								HorizontalContentAlignment="Stretch"
+										Style="{Binding MaterialTextButtonStyle}" BorderThickness="0"
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:31]: Are your branch up to date with latest main ? As we don't have style names starting with "Material" anymore (as I see MaterialTextButtonStyle in your PR for example.
cf. this PR for the previous changes => https://github.com/unoplatform/uno.todo/pull/250/files

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 6]
```diff
@@ -1,6 +1,41 @@
...
+[ReactiveBindable]
 public partial class ExpirationDateViewModel
 {
-    
+	private readonly INavigator _navigator;
+
+	public DateTime DueDate
+	{
+		get;
+		set;//TODO: the idea is return to TaskViewModel with the DueDate value / call BackToTaskMethod
```
Comment (agneszitte-nventive) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:11]: Do we have an issue linked for this TODO ?

**Review by agneszitte-nventive — CHANGES_REQUESTED** [review]

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml [hunk 7]
```diff
@@ -33,114 +102,99 @@
...
+						<utu:AutoLayout Spacing="10"
+										Padding="0,2,0,2"
+										Orientation="Horizontal"
+										utu:AutoLayout.CounterAlignment="Center"
+										utu:AutoLayout.PrimaryAlignment="Stretch">
+							<TextBlock
+									   Text="Today"
+								FontWeight="Bold"
+									   utu:AutoLayout.CounterAlignment="Center"
+									   x:Uid="Task_Details_Add_Due_Date_Label" />
```
Comment (Xiaoy312) [inline src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:127]: 1. So far the naming convention for the x:Uid is NameOfPage_DescriptionOfElement.ElementProperty
2. You should update the resources.l12n.xml and resources.resw files to add the localization.

Comment (Xiaoy312) [reply src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:127]: tooltip

File: src/ToDo/Business/Models/ToDoTaskExtensions.cs [hunk 8]
```diff
@@ -11,4 +11,9 @@ public static ToDoTask WithToggledIsImportant(this ToDoTask task)
 	{
 		return task with { Importance = task.IsImportant ? ToDoTask.TaskImportance.Normal : ToDoTask.TaskImportance.Important };
 	}
+
+	public static ToDoTask WithDeleteDueDate(this ToDoTask task)
```
Comment (Xiaoy312) [inline src/ToDo/Business/Models/ToDoTaskExtensions.cs:15]:
```suggestion
	public static ToDoTask WithDueDate(this ToDoTask task, DateTimeData? dueDate)
	{
	    return task with { DueDateTime = dueDate };
```

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 9]
```diff
@@ -1,6 +1,41 @@
...
+	{
+		_navigator = navigator;
+	}
+
+	private async ValueTask BackToTask()
+	{
+		 await _navigator.NavigateBackWithResultAsync(this, data: new DateTimeData { DateTime = DueDate });
+	}
+
+	public ICommand Today => Command.Async(DoToday);
```
Comment (Xiaoy312) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:23]: Use verb or verb-noun for command, so Select(Today|Tomorrow|NextWeek|CustomDate)

File: src/ToDo.Skia.WPF.Host/ToDo.Skia.Wpf.Host.csproj [hunk 10]
```diff
@@ -8,6 +8,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Identity.Client" Version="4.43.1" />
     <PackageReference Include="Refit" Version="6.3.2" />
+    <PackageReference Include="Uno.Material.WinUI" Version="2.0.0-dev.185" />
```
Comment (Xiaoy312) [inline src/ToDo.Skia.WPF.Host/ToDo.Skia.Wpf.Host.csproj:11]: shouldnt be added to those projects

**Review by Xiaoy312 — COMMENTED** [review]
Couple of things to address:
- conventional commit message
- formatting of xaml in general: ensuring that the are attribute aligned
- if you can, maybe `rebase -i` & `rebase main` to clean up the history a bit

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml (hunk 3 above)
Comment (agneszitte-nventive) [reply src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:21]: cc @kazo0

File: src/ToDo.Skia.WPF.Host/ToDo.Skia.Wpf.Host.csproj (hunk 10 above)
Comment (agneszitte-nventive) [reply src/ToDo.Skia.WPF.Host/ToDo.Skia.Wpf.Host.csproj:11]: yes this change can be removed.
Also `<PackageReference Include="Uno.Material.WinUI" Version="2.0.0-dev.191" />` is already in the app.

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml (hunk 3 above)
Comment (kazo0) [reply src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:21]: Yes, the DatePicker in the flyout does not look like the default Material style. You can see it in the ToDo Figma:
image

File: src/ToDo.UI/App.xaml.host.cs [hunk 11]
```diff
@@ -112,7 +112,7 @@ string ResourceKey(string keyPath)
 			/// Dialogs and Flyouts
 			new ViewMap<AddTaskFlyout, AddTaskViewModel>(),
 			new ViewMap<AddListFlyout, AddListViewModel>(),
-			new ViewMap<ExpirationDateFlyout, ExpirationDateViewModel>(),
+			new ViewMap<ExpirationDateFlyout, ExpirationDateViewModel.BindableExpirationDateViewModel>(),
```
Comment (nickrandolph) [inline src/ToDo.UI/App.xaml.host.cs:115]: This isn't correct - please follow the pattern of the other viewmaps. The BindableExpirationDateViewModel needs to be mapped in the ReactiveClasses.cs.

File: src/ToDo/ToDo.csproj [hunk 12]
```diff
@@ -13,6 +13,7 @@
 		<PackageReference Include="Refit" Version="6.3.2" />
 		<PackageReference Include="Refit.HttpClientFactory" Version="6.3.2" />
 		<PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.17.0" />
+		<PackageReference Include="Uno.Material.WinUI" Version="2.0.0-dev.185" />
```
Comment (nickrandolph) [inline src/ToDo/ToDo.csproj:16]: This shouldn't be here

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs (hunk 6 above)
Comment (ClayAchahui) [reply src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:11]: Not yet..

Comment (ClayAchahui) [reply src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:11]: Actually I'm trying to do that, and the reason why I created this draft PR is to receive suggestion/advice how to approach that

File: src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml (hunk 3 above)
Comment (Xiaoy312) [reply src/ToDo.UI/Views/Dialogs/ExpirationDateFlyout.xaml:21]: I think we should just be using a simple button with icon, text "Pick a Date" and another icon `>` like the previous two.
And, use a not_mobile:DatePickerFlyout or mobile:NativeDatePickerFlyout for this button's Flyout, and bind a starting value.
As for returning a value, you can hook into DatePickerFlyout.DatePicked to handle that. Best would be to create a behavior to wrap the event into a command. OR... simply two-way bind the date, and detect the changes in values.

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs (hunk 6 above)
Comment (Xiaoy312) [reply src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:11]: I think we should just be using a simple button with a not_mobile:DatePickerFlyout or mobile:NativeDatePickerFlyout for flyout, and bind a starting value.
As for returning a value, you can hook into DatePickerFlyout.DatePicked to handle that. Best would be to create a behavior to wrap the event into a command. OR... simply two-way bind the date, and detect the changes in values.

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 13]
```diff
@@ -1,6 +1,41 @@
 namespace ToDo.Presentation.Dialogs;
 
+[ReactiveBindable]
```
Comment (dr1rrb) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:3]: This is no longer needed

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs (hunk 6 above)
Comment (dr1rrb) [reply src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:11]: For now the the easiest solution will indeed to use a `DatePicker` and data bind two-way the value to a `DueDate` property. However, you should use a state for that, and then use the `State.Execute()` to automatically navigate back when a date has been chosen (cf. suggestions below).

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 14]
```diff
@@ -1,6 +1,41 @@
...
+
+	public DateTime DueDate
+	{
+		get;
+		set;//TODO: the idea is return to TaskViewModel with the DueDate value / call BackToTaskMethod
+	}
+	private ExpirationDateViewModel(INavigator navigator)
+	{
+		_navigator = navigator;
+	}
```
Comment (dr1rrb) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:16]:
```suggestion
		DueDate.Execute(async (date, ct) => await _navigator.NavigateBackWithResultAsync(this, data: new DateTimeData { DateTime = date }, cancellation: ct);
	}
```

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 15]
```diff
@@ -1,6 +1,41 @@
...
+
+	private async ValueTask BackToTask()
+	{
+		 await _navigator.NavigateBackWithResultAsync(this, data: new DateTimeData { DateTime = DueDate });
+	}
```
Comment (dr1rrb) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:21]: This is probably not use anymore
```suggestion
```

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 16]
```diff
@@ -1,6 +1,41 @@
...
+
+	public ICommand SelectToday => Command.Async(DoSelectToday);
+	private async ValueTask DoSelectToday(CancellationToken ct)
```
Comment (dr1rrb) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:24]: You no longer need to explicitly declare the `ICommand` when your method is public and has at most 1 parameter (the `CommandParameter`) + an optional cancellation token.
```suggestion
	public async ValueTask SelectToday(CancellationToken ct)
```

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 17]
```diff
@@ -1,6 +1,41 @@
...
+	{
+		await _navigator.NavigateBackWithResultAsync(this, data: new DateTimeData { DateTime = DateTime.Today }, cancellation: ct);
+	}
+
+	public ICommand SelectTomorrow => Command.Async(DoSelectTomorrow);
+	private async ValueTask DoSelectTomorrow(CancellationToken ct)
```
Comment (dr1rrb) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:30]:
```suggestion
	public async ValueTask SelectTomorrow(CancellationToken ct)
```

File: src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs [hunk 18]
```diff
@@ -1,6 +1,41 @@
...
+	{
+		await _navigator.NavigateBackWithResultAsync(this, data: new DateTimeData { DateTime = DateTime.Today.AddDays(1) }, cancellation: ct);
+	}
+
+	public ICommand SelectNextWeek => Command.Async(DoSelectNextWeek);
+	private async ValueTask DoSelectNextWeek(CancellationToken ct)
```
Comment (dr1rrb) [inline src/ToDo/Presentation/Dialogs/ExpirationDateViewModel.cs:36]:
```suggestion
	public async ValueTask SelectNextWeek(CancellationToken ct)
```

File: src/ToDo/Presentation/TaskViewModel.cs [hunk 19]
```diff
@@ -43,4 +47,24 @@ private async ValueTask DoDelete(ToDoTask task, CancellationToken ct)
...
+		{
+			return;
+		}
+
+		var result = await response.Result;
+
+		var date = result.SomeOrDefault()?.DateTime;
+		if (date is not null)
+		{
+			var updated = task with { DueDateTime = (task.DueDateTime ?? new()) with { DateTime = (DateTime)date, TimeZone="UTC" } };
```
Comment (dr1rrb) [inline src/ToDo/Presentation/TaskViewModel.cs:65]: Why are setting the `TimeZone` to "UTC" ? I would assume it should be local ... and actually if possible I would not specify it.

**Pull request closed without merging**